src/piadm.sh

```diff
@@ -195,6 +196,107 @@ DEFAULT_URL_PREFIX=https://us-central.manta.mnx.io/Joyent_Dev/public/SmartOS/
+#   	 is expected to be the last path component of the URL,
+#   	 or the literal string "latest".
+#
+# Environment Variables:
+#   PIADM_NO_SUM     - If set to 1, skips checksum validation.
+#   PIADM_SUM_URL    - If set, overrides the default URL used
+#                      the checksum file used for platform images.
+#
+# Notes:
+#   - If "latest" is used, up to 1024 PIs will be listed from ${URL_PREFIX}.
+#     The last one is assumed to be the latest.
+#   - On error, this function exits with code 1, following PIADM(8)
+#     convention: the error indicates a failure, but no changes were made
+#     to the system.
+declare csum_platform=""
+declare stamp=""
+fetch_csum() {
+	if [[ -z "$csum_platform" ]]; then
+		local platform_file
+		if [ "$1" != "latest" ]; then
+			IFS='/' read -ra array <<< "$1"
+			platform_file="${array[-1]}"
+			stamp="${array[-2]}"
+		else
+			#latest is a special case, we need find out stamp
+			stamps_url="${URL_PREFIX%/}?limit=1024"
+			stamp=$("${CURL[@]}" "${stamps_url}" |\
+				json -ga -c "this.name.match(/Z$/)" | json -ag name |\
+				sort  | tail -1; exit "${PIPESTATUS[0]}")
+			code=$?
+			if [[ $code -ne 0 ]]; then
+				eecho "Curl failed fetching PI data from ${stamps_url}"
+				return 1
+			fi
+			platform_file="smartos-${stamp}.iso"
+		fi
+		if [[ -z ${PIADM_SUM_URL} ]];then
+			local csum_url="${URL_PREFIX}${stamp}/${PIADM_DIGEST_ALGORITHM:-md5}sums.txt"
+		else
+			local csum_url="${PIADM_SUM_URL}"
+		fi
+		csum_platform=$("${CURL[@]}"  "${csum_url}" |\
+			awk -v pattern="${platform_file}"\
+			'$0 ~ pattern { print $1; exit}'; exit "${PIPESTATUS[0]}")
+		code=$?
+		if [[ $code -ne 0 ]]; then
+			eecho "fetching checksums from ${csum_url}"
+			#force re-calculation of checksum on the next call.
+			csum_platform=""
+			stamp=""
+			return 1
+		fi
+		echo "${csum_platform}"
+		return 0
+	else
+		echo "${csum_platform}"
+		return 0
+	fi
+}
+
+# validate_csum
+#
+# $1 is the URL from where the PI was downloaded.
+# $2 is the on disk PI image that was download from $1.
+#
+# Error code will always be 1, following the PIADM(8) convention where
+# a return code of 1 means: an error has occurred, but no change was made.
+# Environment variable PIADM_DIGEST_ALGORITHM controls the checksum
+# algorithm used by digest(1), by default md5 is used.  
+validate_csum() {
+	if [[ $PIADM_NO_SUM -eq 1 ]]; then
+		vecho "WARNING: Not using validation checksum"
+		return 0
+	fi
+	published_csum=$(fetch_csum "$1")
+	code=$?
+	if [[ $code -ne 0 ]]; then
+		eecho "Could not get checksum for PI exit code: ${code}"
+		return 1
+	fi
+	local_csum=$(digest -a "${PIADM_DIGEST_ALGORITHM:-md5}" "${2}")
+	code=$?
+	if [[ $code -ne 0 ]]; then
+		eecho "checksum failed for $2, algorithm used: ${PIADM_DIGEST_ALGORITHM}"
+		return 1
+	fi
+	if [[ "${published_csum}" != "${local_csum}" ]]; then
+		vecho "published_checksum: ${published_csum}"
+		vecho "local_checksum:     ${local_csum}"
```

Comment:
I'd move these vecho OUTSIDE the inequality check before it.  If I'm using -v I wanna see these anyway.